src/hotspot/share/opto/compile.cpp

```diff
@@ -4199,11 +4199,25 @@ bool Compile::needs_clinit_barrier(ciInstanceKlass* holder, ciMethod* accessing_
 //------------------------------verify_bidirectional_edges---------------------
 // For each input edge to a node (ie - for each Use-Def edge), verify that
 // there is a corresponding Def-Use edge.
-void Compile::verify_bidirectional_edges(Unique_Node_List &visited) {
+void Compile::verify_bidirectional_edges(Unique_Node_List& visited, const Unique_Node_List* root_and_safepoints) const {
   // Allocate stack of size C->live_nodes()/16 to avoid frequent realloc
   uint stack_size = live_nodes() >> 4;
-  Node_List nstack(MAX2(stack_size, (uint)OptoNodeListSize));
-  nstack.push(_root);
+  Node_List nstack(MAX2(stack_size, (uint) OptoNodeListSize));
+  if (root_and_safepoints != nullptr) {
```

Comment:
Can you say in which cases we don't have `root_and_safepoints`? Why is it ok not to also start at SafePoint in those cases?

Reply:
I think you should also say that we start the traversal from Root and Safepoints, just like during CCP.

Reply:
I've improved the comment on the declaration of `verify_graph_edges`: it is the only caller of `verify_bidirectional_edges`, which acts more like a helper, and `verify_graph_edges` is the one called a bit everywhere. Also, I think it's not an implementation detail, but a signature/contract thing: when writing a call to `verify_graph_edges`, I must know what I need to provide in `root_and_safepoints`, or when I can omit it. So now, I hope it's documented.